silkworm/db/snapshots/index/btree_index.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+   Copyright 2024 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "btree_index.hpp"
+
+#include <utility>
+
+#include <gsl/util>
+
+#include <silkworm/core/common/endian.hpp>
+#include <silkworm/core/common/util.hpp>
+#include <silkworm/infra/common/ensure.hpp>
+
+namespace silkworm::snapshots::index {
+
+BTreeIndex::BTreeIndex(seg::Decompressor& kv_decompressor,
+                       std::filesystem::path index_file_path,
+                       std::optional<MemoryMappedRegion> index_region,
+                       uint64_t btree_fanout)
+    : file_path_(std::move(index_file_path)) {
+    ensure(kv_decompressor.is_open(), "BTreeIndex: KV file decompressor must be opened");
+
+    // Gracefully handle the case of empty index file before memory mapping to avoid error
+    if (std::filesystem::file_size(file_path_) == 0) {
+        return;
+    }
+
+    // Either use given memory-mapped region or create a new one
+    memory_file_ = std::make_unique<MemoryMappedFile>(file_path_, index_region);
+    SILKWORM_ASSERT(memory_file_->size() > 0);
+    const auto memory_mapped_range = memory_file_->region();
+
+    // Read encoded Elias-Fano 32-bit list of integers representing data offsets
+    data_offsets_ = EliasFanoList32::from_encoded_data(memory_mapped_range);
+    ensure(data_offsets_->sequence_length() > 0, "BTreeIndex: invalid zero-length data offsets");
+
+    const auto encoded_nodes = memory_mapped_range.subspan(data_offsets_->encoded_data_size());
+
+    // Let the OS know we're going to read data sequentially now, then restore normal (i.e. unknown) reading behavior
+    kv_decompressor.advise_sequential();
+    [[maybe_unused]] auto _ = gsl::finally([&]() { kv_decompressor.advise_normal(); });
```

Comment:
feels unnecessary
there's a ReadModeGuard inside begin() that switches to ReadMode::kSequential temporarily